src/messaging/ExchangeContext.cpp

```diff
@@ -109,35 +110,34 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
     if (sendFlags.Has(SendMessageFlags::kExpectResponse))
     {
         // Only one 'response expected' message can be outstanding at a time.
-        VerifyOrExit(!IsResponseExpected(), err = CHIP_ERROR_INCORRECT_STATE);
+        if (IsResponseExpected())
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
```

Comment:
This seems to be fixing a bug where if we had a "response expected" message pending and tried to send another one we would fail to send the new one _and_ stop waiting for the old one.

That's a good thing to fix, but could you please add a test for this situation?  Followup is OK, as long as we check that the test fails without this PR and passes with it.

Also, it might be better to fix it as a separate PR, to keep this refactoring PR behavior-preserving....

Reply:
Added a comment for it